### PR TITLE
feat(connector): Refactor `metrics, source_info` into `source_ctx` and add `ErrorSuppressor` for user source errors

### PR DESCRIPTION
**Pull request by jon-chuang**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
Refactor `metrics, source_info` into `source_ctx` and add `ErrorSuppressor` for user source errors

## Checklist For Contributors

- [X] I have written necessary rustdoc comments
- [X] I have added necessary unit tests and integration tests
- [X] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features).
- [X] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [X] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [X] My PR **DOES NOT** contain user-facing changes.

### Related
Part of: https://github.com/risingwavelabs/risingwave/issues/7831

Generating errors:

In RisingWave:
```sql
create table s1 (v1 int, v2 varchar) with (
  connector = 'kafka',
  topic = 'kafka_1_partition_topic',
  properties.bootstrap.server = '127.0.0.1:29092',
  scan.startup.mode = 'earliest'
) row format json;
```

Ingest data to kafka:
```python3
from kafka import KafkaProducer
import string
import random

producer = KafkaProducer(bootstrap_servers='127.0.0.1:29092')

for _ in range(1000):
  rstr = ''.join(random.choices(string.ascii_lowercase, k=5))
  mystr = f'{{{rstr}: 1}}'
  print(mystr)
  producer.send("kafka_1_partition_topic", value=mystr.encode())
```

We see that they are suppressed:
image

### CODE REVIEW

File: src/connector/src/source/base.rs [hunk 1]
```diff
@@ -76,66 +77,80 @@ pub trait SplitEnumerator: Sized {
...
     }
 
-    pub(crate) fn for_test() -> Self {
-        Self {
-            metrics: Arc::new(SourceMetrics::unused()),
-            table_id: u32::MAX,
-            fragment_id: u32::MAX,
-        }
-    }
+    // pub(crate) fn for_test() -> Self {
```
Comment (jon-chuang) [inline src/connector/src/source/base.rs:135]: remove

File: src/connector/src/source/base.rs [hunk 2]
```diff
@@ -21,10 +21,11 @@ use bytes::Bytes;
 use enum_as_inner::EnumAsInner;
 use futures::stream::BoxStream;
 use itertools::Itertools;
+use parking_lot::Mutex;
```
Comment (fuyufjh) [inline src/connector/src/source/base.rs:24]: Not a real problem but why not `tokio::sync::Mutex`? 👀

**Review by fuyufjh — COMMENTED** [review]
LGTM.

By the way, is it possible to show a table in Grafana instead of curve?

**Review by fuyufjh — APPROVED** [review]

File: src/connector/src/source/base.rs (hunk 2 above)
Comment (jon-chuang) [reply src/connector/src/source/base.rs:24]: No issue, but we use `parking_lot` elsewhere. 

Realistically, the mutex is per-actor, so it should not be a performance bottleneck.

I think either is fine.

**Review by tabVersion — APPROVED** [review]
LGTM

**Comment by jon-chuang** [comment]
> By the way, is it possible to show a table in Grafana instead of line chart?

Sure, I will try this out.

**Pull request merged**